internal/api/impl/v1/dashboard/validator/validator.go

```diff
@@ -0,0 +1,220 @@
+	data, err := os.ReadFile(baseDefPath)
+	if err != nil {
+		logrus.WithError(err).Errorf("Not able to read the base panel definition file at %s, shutting down..", baseDefPath)
+	}
+	baseDef := v.context.CompileBytes(data)
+
+	// initial load of plugins schemas
+	pluginsPath := fmt.Sprintf("%s/%s", v.schemasConf.Path, pluginsFolderName)
+	schemas, err := loadSchemas(v.context, baseDef, pluginsPath)
+	if err != nil {
+		logrus.WithError(err).Error("Not able to retrieve the list of schema files")
+	}
+
+	v.baseDef = baseDef
+	v.schemas = schemas
+}
+
+/*
+ * Validate a dashboard.
+ * The dashboard's panels are matched against the known list of CUE definitions (schemas).
+ * If no schema matches, the validation fails.
+ */
+func (v *validator) Process(dashboard *v1.Dashboard) error {
+	logrus.Tracef("Dashboard to validate: %+v", dashboard)
+
+	// init the validator if it wasn't already
+	if s, _ := v.baseDef.String(); len(s) == 0 {
+		v.initSchemas()
+	}
+
+	var res error
+
+	// go through the panels list
+	// the processing stops as soon as it detects an invalid panel TODO: can probably be improved
+	for k, panel := range dashboard.Spec.Panels {
+		logrus.Tracef("Panel to validate: %s", string(panel))
+
+		// compile the JSON panel into a CUE Value
+		value := v.context.CompileBytes(panel)
+
+		// retrieve panel's kind
+		kind, err := value.LookupPath(cue.ParsePath("kind")).String()
+		if err != nil {
+			logrus.Error("This panel doesn't contain the required \"kind\" field")
+			res = err
+			break
+		}
+
+		// retrieve the corresponding schema
+		schema, ok := v.schemas.Load(kind)
+		if !ok {
+			logrus.Errorf("%s is not valid panel: unknown kind %s", k, kind)
+			res = errors.New("Unknown panel kind")
+			break
+		}
+		logrus.Tracef("Matching panel %s against schema: %+v", k, schema)
+
+		// do the validation
+		unified := value.Unify(schema.(cue.Value))
+		opts := []cue.Option{
+			cue.Concrete(true),
+			cue.Attributes(true),
+			cue.Definitions(true),
+			cue.Hidden(true),
+		}
+		err = unified.Validate(opts...)
+		if err != nil {
+			logrus.Errorf("%s is not a valid %s panel", k, kind)
+			res = err
+			break
+		}
+	}
+
+	if res == nil {
+		logrus.Debug("This dashboard is OK, all its panels are valid")
+	} else {
+		logrus.WithError(res).Error("This dashboard is KO, at least 1 panel is invalid")
+	}
+
+	return res
+}
+
+/*
+ * Load the known list of schemas into the validator
+ */
+func (v *validator) LoadSchemas(path string) {
+	schemas, err := loadSchemas(v.context, v.baseDef, fmt.Sprintf("%s/%s", path, pluginsFolderName))
+	if err != nil {
+		logrus.WithError(err).Error("Not able to retrieve the list of schema files")
+		return
+	}
+
+	v.schemas = schemas
+	logrus.Info("Schemas list (re)loaded")
+}
+
+/*
+ * Load & return the known list of schemas
+ */
+func loadSchemas(context *cue.Context, baseDef cue.Value, path string) (*sync.Map, error) {
+	var schemas sync.Map
+
+	files, err := ioutil.ReadDir(path)
```

Comment:
as state by the go doc, you should use `os.ReadDir`: https://pkg.go.dev/io/ioutil#ReadDir

> As of Go 1.16, os.ReadDir is a more efficient and correct choice: it returns a list of fs.DirEntry instead of fs.FileInfo, and it returns partial results in the case of an error midway through reading a directory.